File: backend/src/main/java/lan/dk/podcastserver/manager/worker/rss/RSSFinder.java
package lan.dk.podcastserver.manager.worker.rss;

import com.github.davinkevin.podcastserver.service.ImageService;
import io.vavr.control.Option;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.manager.worker.Finder;
import com.github.davinkevin.podcastserver.service.JdomService;
import lombok.RequiredArgsConstructor;
import org.jdom2.Element;
import org.springframework.stereotype.Service;

import javax.validation.constraints.NotEmpty;
import java.util.Objects;

import static io.vavr.API.Option;

/**
 * Created by kevin on 22/02/15
 */
@Service("RSSFinder")
@RequiredArgsConstructor
public class RSSFinder implements Finder {

    private static final String CHANNEL = "channel";
    private static final String TITLE = "title";
    private static final String DESCRIPTION = "description";
    private static final String IMAGE = "image";
    private static final String URL = "url";
    private static final String HREF = "href";

    private final JdomService jdomService;
    private final ImageService imageService;

    @Override
    public Podcast find(String url) {
        return jdomService
                .parse(url)
                .map(x -> x.getRootElement().getChild(CHANNEL))
                .filter(Objects::nonNull)
                .map(e -> this.xmlToPodcast(e, url))
                .getOrElse(Podcast.DEFAULT_PODCAST);
    }

    private Podcast xmlToPodcast(Element element, String url) {
        return Podcast.builder()
                .type("RSS")
                .url(url)
                .title(element.getChildText(TITLE))
                .description(element.getChildText(DESCRIPTION))
                .cover(coverUrlOf(element))
                .build();
    }

    private Cover coverUrlOf(Element channelElement) {
        return getRssImage(channelElement)
                .orElse(getItunesImage(channelElement))
                .map(imageService::getCoverFromURL)
                .getOrElse(Cover.DEFAULT_COVER);
    }

    private Option<String> getItunesImage(Element channelElement) {
        return Option(channelElement.getChild(IMAGE, JdomService.Companion.getITUNES_NAMESPACE()))
                .map(itunesImage -> itunesImage.getAttributeValue(HREF));
    }

    private Option<String> getRssImage(Element channelElement) {
        return Option(channelElement.getChild(IMAGE))
                .map(rssImage -> rssImage.getChildText(URL));
    }

    public Integer compatibility(@NotEmpty String url) {
        return Integer.MAX_VALUE-1;
    }

}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/find/finders/rss/RSSFinder.kt
package com.github.davinkevin.podcastserver.find.finders.rss

import com.github.davinkevin.podcastserver.find.FindCoverInformation
import com.github.davinkevin.podcastserver.find.FindPodcastInformation
import com.github.davinkevin.podcastserver.find.finders.Finder
import com.github.davinkevin.podcastserver.find.finders.fetchFindCoverInformation
import com.github.davinkevin.podcastserver.service.image.ImageService
import org.jdom2.Element
import org.jdom2.Namespace
import org.jdom2.input.SAXBuilder
import org.springframework.web.client.RestClient
import org.springframework.web.client.body
import java.io.StringReader
import java.net.URI

/**
 * Created by kevin on 22/02/15
 */
class RSSFinder(
        private val imageService: ImageService,
        private val rcb: RestClient.Builder
) : Finder {

    private val itunesNS = Namespace.getNamespace("itunes", "http://www.itunes.com/dtds/podcast-1.0.dtd")!!

    override fun findPodcastInformation(url: String): FindPodcastInformation? {
        val content = rcb
            .clone()
            .baseUrl(url)
            .build()
            .get()
            .retrieve()
            .body<String>()
            ?: return null

        val channel = SAXBuilder().build(StringReader(content))
            .rootElement
            .getChild("channel")

        val cover = findCover(channel)

        return FindPodcastInformation(
            type = "RSS",
            url = URI(url),
            title = channel.getChildText("title"),
            description = channel.getChildText("description"),
            cover = cover
        )
    }

    private fun findCover(channelElement: Element): FindCoverInformation? {
        val rss = channelElement.getChild("image")?.getChildText("url")
        val itunes = channelElement.getChild("image", itunesNS)?.getAttributeValue("href")

        val url = rss ?: itunes ?: return null

        return imageService.fetchFindCoverInformation(URI(url))
    }

    override fun compatibility(url: String): Int = Integer.MAX_VALUE - 1
}
